### PR TITLE
Debugging in Photon.js

### PR DESCRIPTION
**Pull request by timsuchanek**

### CODE REVIEW

File: photonjs/README.md [hunk 1]
```diff
@@ -161,6 +162,44 @@ type Profile = {
...
+```ts
+
+// Either like this
+const photon = new Photon({
+  logQueries: 'pipe',
+})
+
+// Or like this
+const photon = new Photon({
+  debug: {
```
Comment (schickling) [inline photonjs/README.md]: Is it also possible to provide `debug: true`?

File: photonjs/README.md [hunk 2]
```diff
@@ -161,6 +162,44 @@ type Profile = {
...
+    queries: 'pipe', // | true | 'inherit', 'inherit' = true
+    engine: 'pipe'
```
Comment (schickling) [inline photonjs/README.md]: What's the difference between `queries` and `engine`?

File: photonjs/README.md (hunk 1 above)
Comment (schickling) [inline photonjs/README.md]: Not sure whether `debug` is the right term. To me `log` or `logging` sounds more appropriate?

Comment (timsuchanek) [reply photonjs/README.md]: Just moved it out of `debug`.

File: photonjs/README.md (hunk 2 above)
Comment (timsuchanek) [reply photonjs/README.md]: Moved everything `engine` related out to `log`.

File: photonjs/README.md (hunk 1 above)
Comment (timsuchanek) [reply photonjs/README.md]: Totally. New suggestion: Calling it `log`.

File: photonjs/README.md [hunk 3]
```diff
@@ -162,6 +166,131 @@ type Profile = {
...
+### Logging
+
+All engine related logs can be configured with the `log` property.
+These are examples how to specify different log levels:
+
+Just providing the log levels, stdout as default
+
+```ts
+const photon = new Photon({
+  log: ['info', 'query'],
```
Comment (schickling) [inline photonjs/README.md]: This seems to be inconsistent with `export type LogLevel = 'INFO' | 'WARN' | 'QUERY'`.

File: photonjs/README.md [hunk 4]
```diff
@@ -162,6 +166,131 @@ type Profile = {
...
+      cwd?: string
+      binaryPath?: string
+    }
+  }
+}
+```
+
+## Debugging & Logging
+
+### Debugging
```
Comment (schickling) [inline photonjs/README.md]: Please switch the order of debugging and logging in the spec. Logging is user facing and should come first, debugging is more low level and used less often.

File: photonjs/README.md [hunk 5]
```diff
@@ -162,6 +166,131 @@ type Profile = {
...
+    hooks?: Hooks
+    engine?: {
+      cwd?: string
+      binaryPath?: string
+    }
+  }
+}
+```
+
+## Debugging & Logging
```
Comment (schickling) [inline photonjs/README.md]:
```suggestion
## Logging & Debugging
```

File: photonjs/README.md [hunk 6]
```diff
@@ -162,6 +166,131 @@ type Profile = {
...
+})
+```
+
+Changing on a per log level, where the logs end up: As an event or in stdout.
+
+```ts
+const photon = new Photon({
+  log: [
+    {
+      level: 'info',
```
Comment (schickling) [inline photonjs/README.md]: This seems to be inconsistent with `export type LogLevel = 'INFO' | 'WARN' | 'QUERY'`.

File: photonjs/README.md [hunk 7]
```diff
@@ -350,16 +479,16 @@ photon.user.find({
...
+| Operation             | Touches Record | Touches Connection |
+| --------------------- | -------------- | ------------------ |
+| `connect`             | No             | Yes                |
+| `disconnect`          | No             | Yes                |
+| ??? `resetAndConnect` | No             | Yes                |
+| `connectOrCreate`     | Yes            | Yes                |
+| `create`              | Yes            | Yes                |
+| `update`              | Yes            | No                 |
+| `replace`             | Yes            | No                 |
+| `delete`              | Yes            | Yes                |
```
Comment (schickling) [inline photonjs/README.md]: This seems unrelated to the active PR. Please undo these changes.

File: photonjs/README.md [hunk 8]
```diff
@@ -527,6 +656,32 @@ await photon.user.find('bobs-id').update({
...
+  firstName: 'Alice',
+  lastName: 'Doe',
+  email: 'alice@prisma.io',
+  profile: { imageUrl: 'http://...', imageSize: 100 },
+  posts: e => [
+    e.create({ title: 'New post', body: 'Hello world', published: true }),
+    e.create({ title: 'New post 2', body: 'Hello world 2', published: true })
+  ]
+})
+
```
Comment (schickling) [inline photonjs/README.md]: This seems unrelated to the active PR. Please undo these changes.

File: photonjs/README.md [hunk 9]
```diff
@@ -776,6 +931,7 @@ type DynamicResult4 = {
   key: string
   records: User[]
 }
+
```
Comment (schickling) [inline photonjs/README.md]: This seems unrelated to the active PR. Please undo these changes.

File: photonjs/README.md [hunk 10]
```diff
@@ -804,8 +960,12 @@ await photon.user.findMany().group({ by: u => u.email.toLowerCase() })
...
 
 const values: string[] = await photon.post
-  .findMany()
-  .distinct({ title: true })
+  .findMany({
+    distinct: {
+      title: true,
+    },
+  })
+  // .distinct({ title: true })
```
Comment (schickling) [inline photonjs/README.md]: This seems unrelated to the active PR. Please undo these changes.

File: photonjs/README.md [hunk 11]
```diff
@@ -900,16 +1060,29 @@ Note: Combined with OCC (i.e. `if`) also known as "unit of work"
...
+      user: {
+        if: {
+          id: '',
+        },
+      },
+    },
+    m2,
+  ],
+  { transaction: true },
+)
```
Comment (schickling) [inline photonjs/README.md]: This seems unrelated to the active PR. Please undo these changes.

File: photonjs/README.md [hunk 12]
```diff
@@ -1300,14 +1473,15 @@ To solve these two use case, Photon can do the following:
...
+- [ ] Shortcut for lists: `tags: [""]` instead of `tags: {set: [""]}`
 - [ ] Decouple engine `connect` API from Photon instance (solves: https://github.com/prisma/photonjs/issues/153)
 - [ ] Should we load data by default for create operations?
 - [ ] "Dataloader"
 - [ ] Query engine logging
 - [x] Distinct
 - [ ] Tracing
 - [ ] `find` vs `findUnique` vs `get` ...
-- [x] Terminology: link vs connect (https://github.com/prisma/photonjs/issues/227 and https://github.com/prisma/specs/issues/140#issuecomment-530821669)
+- [ ] Terminology: link vs connect (https://github.com/prisma/photonjs/issues/227 and https://github.com/prisma/specs/issues/140#issuecomment-530821669)
```
Comment (schickling) [inline photonjs/README.md]: This seems unrelated to the active PR. Please undo these changes.

File: photonjs/README.md [hunk 13]
```diff
@@ -1300,14 +1473,15 @@ To solve these two use case, Photon can do the following:
 
 ### Small & self-contained
 
+- [ ] Shortcut for lists: `tags: [""]` instead of `tags: {set: [""]}`
```
Comment (schickling) [inline photonjs/README.md]: This seems unrelated to the active PR. Please undo these changes.

File: photonjs/README.md [hunk 14]
```diff
@@ -1340,3 +1514,9 @@ To solve these two use case, Photon can do the following:
   - [ ] How to query records that were "touched" during nested writes
   - [ ] (Nested) page info
 - [ ] Union queries
+
+```ts
+import { Post } from '...'
+
+const x = photon.raw<Pick<Post, 'id'>>('...')
+```
```
Comment (schickling) [inline photonjs/README.md]: This seems unrelated to the active PR. Please undo these changes.

**Comment by timsuchanek** [comment]
Closed in favor of https://github.com/prisma/specs/pull/307
